### PR TITLE
Generate toc for dotnet

### PR DESCRIPTION
**Pull request by sima-zhu**
# Contributing to the Azure SDK
This is the PR to generate toc for dotnet pipeline.

Here is the testing pipeline: https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1977544&view=results
Here is the toc looks like: https://github.com/Azure/azure-docs-sdk-dotnet/blob/daily/2022-11-09/docs-ref-toc/reference-unified.yml
Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

### CODE REVIEW

File: eng/scripts/Language-Settings.ps1
Comment (sima-zhu) [inline eng/scripts/Language-Settings.ps1:408]: Some of the management track 1 packages do not exist in public feeds anymore. 
image

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [inline eng/scripts/docs/Docs-ToC.ps1:29]: We setup `PackageSourceOverride` variable at `Update Docs Onboarding for Daily branch` step

File: eng/scripts/Language-Settings.ps1
Comment (sima-zhu) [reply eng/scripts/Language-Settings.ps1:408]: Use the try catch to skip the failures

File: eng/pipelines/docindex.yml
Comment (weshaggard) [inline eng/pipelines/docindex.yml:9]: Do we need to add the word Compliant at the end here?

File: eng/pipelines/docindex.yml
Comment (weshaggard) [inline eng/pipelines/docindex.yml:76]: nit: spacing
```suggestion

      - pwsh: |
          $publicFeedUrl = "https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json"
          Write-Host "##vso[task.setvariable variable=PackageSourceOverride]$publicFeedUrl"
        displayName: Set package source variable
        workingDirectory: $(DocRepoLocation)
        
```

Is there a reason this needs to be set as a variable dynamically here as opposed to hardcode below?

File: eng/pipelines/docindex.yml
Comment (weshaggard) [inline eng/pipelines/docindex.yml:106]: Are we certain we want to force push now?

File: eng/scripts/Language-Settings.ps1
Comment (weshaggard) [reply eng/scripts/Language-Settings.ps1:408]: I suspect they may have never been in our devops feed. If we need namespaces for them we might need to either add them or use nuget.org.

Comment (sima-zhu) [reply eng/scripts/Language-Settings.ps1:408]: It doesn't affect main/live site. Will add some fallback logic here to make sure we download the right package.

File: eng/pipelines/docindex.yml
Comment (sima-zhu) [reply eng/pipelines/docindex.yml:9]: That's what we have in the pool.
image

File: eng/pipelines/docindex.yml
Comment (sima-zhu) [reply eng/pipelines/docindex.yml:76]: It is needed for toc generation. That's how we know whether to use public feeds for daily docs or not. 

Our common script doesn't provide the $PackageSourceOverride parameters, as other languages have their way to use this.
Java: we have a step to configure .m2 settings. [Link](https://github.com/Azure/azure-sdk-for-java/blob/main/eng/pipelines/docindex.yml#L96)

.NET has no way to pass this to the agent. So use variable for .NET only and read variable from script.

File: eng/pipelines/docindex.yml
Comment (sima-zhu) [reply eng/pipelines/docindex.yml:106]: We might have nightly pipeline pushes commits during docindex run, so it complains about the commits behind sometime.

By force pushing the commit, we might lose one or two commits from nightly pipeline. The miss on the commits is acceptable to me as following reasons.  First of all, it doesn't occur very often. Second, it doesn't always affect the same package. Last, this is daily refresh docs, so it will show up the right thing in the next day.

File: eng/pipelines/docindex.yml
Comment (sima-zhu) [reply eng/pipelines/docindex.yml:77]: Here is the place to use the variable.
https://github.com/Azure/azure-sdk-for-net/pull/32062/files#r1008278055

File: eng/pipelines/docindex.yml
Comment (weshaggard) [reply eng/pipelines/docindex.yml:9]: @mikeharder do you think we should stick with the image name we have configured in the pool? Or since we are only allowed one image per pool should we use a general name like "windows" for the vmImage name in our yml files?

File: eng/pipelines/docindex.yml
Comment (weshaggard) [reply eng/pipelines/docindex.yml:77]: Ok so is more about setting this as an env variable. Can you please add a comment to that effect here so we don't accidently remove it later and break that usage.

File: eng/pipelines/docindex.yml
Comment (weshaggard) [reply eng/pipelines/docindex.yml:108]: As long as we don't end up overwriting any manual commits, we might have that I suppose this will be ok.

File: eng/pipelines/docindex.yml
Comment (weshaggard) [reply eng/pipelines/docindex.yml:9]: @sima-zhu in the meantime yes keep the name matching the image.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:33]: Should we use the custom feed as the first part of this string? As it stands it seems like we are hardcoding this and it will need to be in sync with the other location we are setting. 'https://pkgs.dev.azure.com/azure-sdk/public/_packaging/azure-sdk-for-net/nuget/v3/index.json' If we ever pass in another value there then this will be broken.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:21]: We should add some tests for these new functions.

File: eng/pipelines/docindex.yml
Comment (sima-zhu) [reply eng/pipelines/docindex.yml:9]: I have this `vmImage` as convention. Based on my testing, it is ok to use pool name only without specifying `vmImage` name.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:33]: The location we hard-coded before can't use for download. I did not find how we utilize the public feeds for downloading purpose.

We can add download-devOpsPackages in `Invoke-DevOps.ps1` so we do not hardcoded anymore. A little extra work as the changes are inside of common script.

Comment (weshaggard) [reply eng/scripts/docs/Docs-ToC.ps1]: What I want to make sure is that we don't hard code this download path here but we actually pass in another feed in our customPackageSource property. So we need to somehow make those be in-sync.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:30]: I don't think this is correct any longer I believe you want the real nuget feed url instead of the package/version url. YOu should be able to use the same nuget install command with either feed and the command should be the same.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:60]: Part of the value of the nuget directdownload is you also don't need to extract the package as it does it for you already.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [inline eng/scripts/docs/Docs-ToC.ps1:34]: Go ahead with `$package.$version` as folder name since we might need to run both preview and latest. Make the script robust.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:51]:
```suggestion
    Write-Host "Searching for dll file..."
```

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:45]:
```suggestion
    Write-Host "Downloading nupkg package $package with version $version to $tempLocation ...."
```

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (weshaggard) [inline eng/scripts/docs/tests/Toc-Generation-Tests.ps1:3]: Lets see if we can also add some tests for the Toc functions as well.

**Review by weshaggard — COMMENTED** [review]
The changes look reasonable to me but I'd like to wait for @danieljurek to review for final sign-off as he understands the structure of the Toc files better.

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (sima-zhu) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:3]: `Get-Toc-Children` is worth to have more tests as they have some complicated logic. I will leave the `Get-dotnet-DocsMsTocData` as it is since the logic is pretty straight forward and it is the same structure as other languages.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (benbp) [inline eng/scripts/docs/Docs-ToC.ps1:32]: Should we handle `$LASTEXITCODE` for these nuget calls?

File: eng/scripts/docs/Docs-ToC.ps1
Comment (benbp) [inline eng/scripts/docs/Docs-ToC.ps1:58]: Is this supposed to be a `-lt` check? If `-gt` the error message could be a bit confusing.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:32]: Don't want to check the error code in each individual nuget command, so I have error handler when we call the Download function.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:58]: Actually, the dllFile should only has one in each target framework (e.g. netstandard2.0). The ($dllFile.Count -gt 1) is redundant for normal world.

Comment (benbp) [reply eng/scripts/docs/Docs-ToC.ps1:58]: If there are > 1 dll files, and the log warning says "can't find any dll file" I think that could be misleading.

Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:58]: After the test, I found we do have package which includes multiple dlls in one target. Added elseif the the case.
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=1971666&view=logs&j=dc056dfc-c0cf-5958-c8c4-8da4f91a3739&t=98c76ae6-047b-5869-f98c-bcf6099a854a&l=623 
image

```
/tmp/extractNupkg/Azure.Cosmos.4.0.0-preview3/lib/netstandard2.0/Azure.Cosmos.dll
/tmp/extractNupkg/Azure.Cosmos.4.0.0-preview3/lib/netstandard2.0/Microsoft.Azure.Cosmos.Core.dll 
/tmp/extractNupkg/Azure.Cosmos.4.0.0-preview3/lib/netstandard2.0/Microsoft.Azure.Cosmos.Direct.dll 
/tmp/extractNupkg/Azure.Cosmos.4.0.0-preview3/lib/netstandard2.0/Microsoft.Azure.Cosmos.Serialization.HybridRow.dll

```

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:55]: Shouldn't this be $dllFiles as this is a list?

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:67]: You should move this warning after you select the dll so you can include the name of the dll in the output.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:79]: Why are you setting ErrorAction Continue on these Write-Error messages? Also should we should probably be using the LogError helpers in these cases so they output correctly as errors in DevOps.

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (weshaggard) [inline eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: We should have the cosmos packages that were causing issues added as test cases.

Comment (kurtzeborn) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: If cosmos has been problematic, it should definitely be added to your list of test cases, shouldn't it?

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:79]: I have one testing failed in the middle. I switched to logging.ps1 apis if it's non stop behavior.

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (sima-zhu) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: This is newly found case. Will add to test.

Comment (sima-zhu) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: I will use Azure.Cosmos one for testing as this is deterministic. `Microsoft.Azure.Cosmos` will pick the first one which is flaky from test perspective.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1:79]: Tested in local. The tests failed at Log function in logging.ps1 since we did not import the logging script in test script. I switched back the Write-Error

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (weshaggard) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: If we think Microsoft.Azure.Cosmos is flaky that means that we also have a flaky implementation that will give indeterministic results. If that is the case we should fix the implementation to order the dlls so that the result is more stable.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1]: https://github.com/Azure/azure-sdk-for-net/pull/32062/files#diff-4b590925689e1df66d5c28ddc67d15ded4db6d686eb7c7878e0c8238feb57e25R76

I have printed at end. The message here shows the route to find dll

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (sima-zhu) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:40]: Then we need to order the dll first and then fetch the first.

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:31]: It is still worth having the package name and version for diagnostic purposes.

Comment (sima-zhu) [reply eng/scripts/docs/Docs-ToC.ps1]: Actually, we have the $package $version information logged out. No need to log here it is a redundant information. 
https://github.com/Azure/azure-sdk-for-net/pull/32062/files#diff-4b590925689e1df66d5c28ddc67d15ded4db6d686eb7c7878e0c8238feb57e25R45

File: eng/scripts/docs/tests/Toc-Generation-Tests.ps1
Comment (sima-zhu) [reply eng/scripts/docs/tests/Toc-Generation-Tests.ps1:41]: Done with the change

File: eng/scripts/docs/Docs-ToC.ps1
Comment (weshaggard) [inline eng/scripts/docs/Docs-ToC.ps1:80]: Looks like this ErrorAction is leftover and should be removed. Looks like there might be a few of them left over.

**Review by weshaggard — APPROVED** [review]

**Pull request merged**
